docs/source/writing-an-oauthenticator.md

```diff
@@ -0,0 +1,46 @@
+# Writing your own OAuthenticator
+
+There are two ways to write your own OAuthenticator.
+
+## Using GenericOAuthenticator
+
+The first and simplest is to use the `oauthenticator.generic.GenericOAuthenticator` class
+and configuration to set the necessary configuration variables.
+
+- client_id
+- client_secret
+- login_service
+- userdata_url
+- token_url
+- username_key
+
+Example config:
+
+```python
+c.JupyterHub.authenticator_class = "generic"
+
+c.GenericOAuthenticator.oauth_callback_url = 'https://{host}/hub/oauth_callback'
+c.GenericOAuthenticator.client_id = 'OAUTH-CLIENT-ID'
+c.GenericOAuthenticator.client_secret = 'OAUTH-CLIENT-SECRET-KEY'
+c.GenericOAuthenticator.login_service = 'name-of-service-provider'
+c.GenericOAuthenticator.userdata_url = 'url-retrieving-user-data-with-access-token'
+c.GenericOAuthenticator.token_url = 'url-retrieving-access-token-oauth-completion'
+c.GenericOAuthenticator.username_key = 'username-key-for-USERDATA-URL'
+```
+
+Checkout [](getting-started:moodle-setup) and [](getting-started:yandex_setup) for how to configure
+GenericOAuthenticator for Moodle and Yandex.
+
+## Writing your own OAuthenticator class
+
+If you want more advanced features and customization beyond the basics of OAuth,
+you can write your own full OAuthenticator subclass,
+which enables more detailed customization login and logout actions.
+
+The skeleton of an OAuthenticator looks like this:
+
+```{eval-rst}
+.. literalinclude:: example-oauthenticator.py
+```
```

Comment:
I think this can be written directly in MyST markup. I parse the rST code to be invoking a sphinx directive, and that would match in MyST markup as described here: https://myst-parser.readthedocs.io/en/latest/syntax/roles-and-directives.html

So it would become...

````
```{literalinclude} example-oauthenticator.py
```
````

Or perhaps this works?

````
```{literalinclude} example-oauthenticator.py```
````